### ISSUE TITLE
Change title of terminal when ranger executes an executable (e.g. Vim)

### ISSUE THREAD

**Issue by talw** [ranger/ranger#819]
Improvement/feature request

This issue appeared as part of another issue, namely #154.
The issue was closed but the part regarding changing the title of the terminal according to the program used to open the file wasn't addressed apparently.

My usecase is exactly like the one mentioned in #154 
My window manager is configured to change focus to specific programs according to their title.

Even when ranger, say, open a text file in vim, the title of the terminal always remains as at : 
`ranger:<path-to-directory>`

Is this something simple to implement? Any hope of getting this feature in the near future?

Thanks!

**Comment by nfnty** [comment]
This would need to be implemented as a new setting as the current behavior is the one that makes the most sense. PRs are welcome.

**Comment by Vifon** [comment]
I feel it should be handled by either vim or an intermediate script (called from rifle).

**Comment by talw** [comment]
@Vifon
In the case of vim you're absolutely right, I've just discovered it is possible using `set titlescreen=`

Regarding other executables, I'll try using an intermediate script and see how straightforward or not is it.

I doubt it is as straightforward as setting a flag in ranger, but of course I wouldn't want to add superfluous options to ranger :)

**Comment by Vifon** [comment]
A flag in ranger/rifle for displaying a generic title with the used commandline sounds fine too. My main point was that Vim can surely generate a much more descriptive title.

I'll take a look at this in my free time, unless someone will do it before me. :)

**Comment by Vifon** [comment]
@talw After inspecting the rifle code, I think using a wrapper for launching the files will be the most clean solution. rifle is mostly separated from the rest of ranger so no settings from `rc.conf` should be read there.

Here is a temporary (or not?) solution, as a wrapper script:

```python
#!/usr/bin/env python

import curses
import subprocess
import sys

def main(argv=None):
    curses.setupterm()
    commandline = argv[1:]
    title = " ".join(commandline)
    try:
        escaped_title = title.encode('utf-8', 'surrogateescape'). \
            decode('utf-8', 'replace')
        fmt_tup = (
            curses.tigetstr('tsl').decode('latin-1'),
            escaped_title,
            curses.tigetstr('fsl').decode('latin-1'),
        )
    except UnicodeError:
        pass
    else:
        sys.stdout.write("{}rifle: {}{}".format(*fmt_tup))
        sys.stdout.flush()
        subprocess.call(commandline)

if __name__ == '__main__':
    from sys import argv
    main(argv)
```

Call it like you would call sudo or some other "prefix-like" command.

**Comment by talw** [comment]
Thank you kindly for the wrapper script!
I do not have the opportunity to test it now, but I will on the weekend!

If you don't mind me asking, how does this work? The bottom line of this script is that you get:
`sys.stdout.write("{}rifle: {}{}".format(*fmt_tup))`
how will an output to stdout cause the title to change? Or perhaps something else uses that output to change the title? Or I'm supposed to change that `stdout.write()` call with something that sets the title of the terminal?

Thanks again!

**Comment by Vifon** [comment]
I took most of it from here: https://github.com/ranger/ranger/blob/3b30c5eefb0cd088197291602c0b0a1be63dac6e/ranger/gui/ui.py#L373-L385

It outputs an [escape code](https://en.wikipedia.org/wiki/ANSI_escape_code) that will be interpreted by your terminal and should cause it to change its title. This is how it works in most (or should I say "all"?) commandline applications setting the terminal title. The specific escape code is decided by the `curses` library.

You may customize the prefix (`rifle: `) if you want.

**Comment by talw** [comment]
I see, thanks for the script and explanation!
I can close this issue unless someone thinks this shouldn't be closed as long as a flag does not exist.

**Comment by vxcamiloxv** [comment]
I think should be a default behavior like in Midnight Commander, it displays `mc [user@machine]:~/path`

**Comment by Vifon** [comment]
@toonn We can consider it in the future. I'm open to the discussion.

**Comment by tbrodbeck** [comment]
When I switch to my shell (zsh in my case) via S and switch back to ranger (kill the shell), the title is not set back to 'ranger'. Is this the same problem? If not I can open another issue.

**Comment by toonn** [comment]
@tbrodbeck, that's kind of the inverse issue, with ranger being lazy about setting the title after executing a program. Feel free to open a new issue.

**Comment by Federico-Ciuffardi** [comment]
Is it possible to have this behavior on ranger 1.9.3-3 without modifying the source code?

I tried to use the `execute.before` signal but I couldn't find a way to get the  executed command  in the function associated with the signal so I can set the title with it.

**Comment by toonn** [comment]
Yes, take the wrapper script vifon made and change the rifle rules you want this behavior to apply to.
